### PR TITLE
Support for Locally Present Subroutines

### LINKED ISSUE #582: Support for Locally Present Subroutines
**Is your feature request related to a problem?**

COBOL program may call a subroutine that is described in an external file. It would be handy to be able to open this file by calling "Go to Definition" on the subroutine name.

**Describe the solution you'd like**

1. Provide the relative path to the folder with subroutines in the settings (the same way as we have it now for copybooks)
1. In the given example: CALL 'UTIL' USING WS-STUDENT-ID, WS-STUDENT-NAME, the user will be able to click on 'UTIL' and select GoToDefinition to open up the UTIL.cbl file
1. Show the used subroutine in the completion suggestions with the appropriate kind
1. Subroutine usages should present in the "Find All References" list
1. Show a semantic error if the subroutine not found with severity error "WARNING"

**Additional information**

1. May require a new JSON-RPC endpoint to allow the direct client-server connection. 
1. Doesn't change the file content, so should be implemented in the CobolVisitor
1. Requires a new `Occurrences` and `Completion` delegates for subroutines
1. Needs to add a new entity for subroutines to the Use Case Engine in order to test CALL statements automatically

### PR DESCRIPTION
**Pull request by grianbrcom**
Closes #582

### CODE REVIEW

File: clients/cobol-lsp-vscode-extension/src/services/util/FSUtils.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        const localFolderList: string[] = targetFolders
+            .map(getURIFromResource)
+            .filter((url: URL) => url !== undefined)
+            .map((url: URL) => url.href);
+        for (const folder of localFolderList) {
+            let uri: URL = getURIFrom(folder, entityName);
+            if (!uri) {
+                uri = getURIFrom(folder, entityName, extensions);
+            }
+            if (uri) {
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/services/util/FSUtils.ts:80]: if this if (uri) comes first then no need another if, if (!uri)  can be removed

File: clients/cobol-lsp-vscode-extension/src/__tests__/extensionTest.spec.ts [hunk 2]
```diff
@@ -90,6 +90,7 @@ const middleware: Middleware = new Middleware(new CopybookURI(profileService), c
 const languageClientService: LanguageClientService = new LanguageClientService(middleware);
 languageClientService.checkPrerequisites = jest.fn(() => Promise.resolve());
 languageClientService.start = jest.fn().mockResolvedValue(true);
+//languageClientService.addRequestHandler = jest.fn();
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/src/__tests__/extensionTest.spec.ts:93]: Please, remove the commented-out code

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/references/SubroutineLocations.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+import com.broadcom.lsp.cobol.service.CobolDocumentModel;
+import lombok.NonNull;
+import org.eclipse.lsp4j.Location;
+
+import java.util.List;
+import java.util.Map;
+
+/** This class is a provider for locations of subroutine defined/used in the document */
+public class SubroutineLocations implements SemanticLocations {
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/delegates/references/SubroutineLocations.java:25]: I see you miss a test for it. Not a big deal though, but we miss coverage here.

**Review by temanbrcom — COMMENTED** [review]
You didn't add a completion provider for subroutines?

File: server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCaseEngine.java [hunk 4]
```diff
@@ -216,6 +245,8 @@ private static void assertResultEquals(AnalysisResult actual, TestData expected)
     assertResult(
         "Section definition:", expected.getSectionDefinitions(), actual.getSectionDefinitions());
     assertResult("Section usages:", expected.getSectionUsages(), actual.getSectionUsages());
+
+    assertResult("Subroutine usage:", expected.getSubroutineUsages(), actual.getSubroutineUsages());
```
Comment (temanbrcom) [inline server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCaseEngine.java:249]: Should assert also the subroutine definitions

**Review by temanbrcom — CHANGES_REQUESTED** [review]

File: server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCaseEngine.java (hunk 4 above)
Comment (grianbrcom) [reply server/src/test/java/com/broadcom/lsp/cobol/usecases/engine/UseCaseEngine.java:249]: Done

File: server/src/main/java/com/broadcom/lsp/cobol/service/delegates/completions/SubroutineCompletion.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import com.google.inject.Singleton;
+import lombok.NonNull;
+import org.eclipse.lsp4j.CompletionItemKind;
+
+import java.util.Collection;
+
+import static com.broadcom.lsp.cobol.service.delegates.completions.CompletionOrder.SUBROUTINES;
+
+@Singleton
+public class SubroutineCompletion implements Completion {
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/delegates/completions/SubroutineCompletion.java:29]: Missing javadoc

**Review by asatklichov — APPROVED** [review]
🧷

**Comment by grianbrcom** [comment]
Fixed Javadocs and fails on positive tests.

**Review by temanbrcom — APPROVED** [review]
🍾

**Comment by grianbrcom** [comment]
Updated message for missing subroutines.

**Review by temanbrcom — APPROVED** [review]
🚂

**Pull request merged**